File: docs/1.0/core/examples/ruby/queries.rb
require 'chain'

chain = Chain::Client.new
signer = Chain::HSMSigner.new

key = chain.mock_hsm.keys.create
signer.add_key(key, chain.mock_hsm.signer_conn)

chain.assets.create(alias: 'gold', root_xpubs: [key.xpub], quorum: 1)
chain.assets.create(alias: 'silver', root_xpubs: [key.xpub], quorum: 1)
chain.accounts.create(alias: 'alice', tags: {type: 'checking'}, root_xpubs: [key.xpub], quorum: 1)
chain.accounts.create(alias: 'bob', root_xpubs: [key.xpub], quorum: 1)

chain.transactions.submit(signer.sign(chain.transactions.build { |b|
  b.issue asset_alias: 'gold', amount: 1000
  b.issue asset_alias: 'silver', amount: 1000
  b.control_with_account account_alias: 'alice', asset_alias: 'gold', amount: 1000
  b.control_with_account account_alias: 'bob', asset_alias: 'silver', amount: 1000
}))

chain.transactions.submit(signer.sign(chain.transactions.build { |b|
  b.spend_from_account account_alias: 'alice', asset_alias: 'gold', amount: 10
  b.spend_from_account account_alias: 'bob', asset_alias: 'silver', amount: 10
  b.control_with_account account_alias: 'alice', asset_alias: 'silver', amount: 10
  b.control_with_account account_alias: 'bob', asset_alias: 'gold', amount: 10
}))

chain.assets.create(alias: 'bank1_usd_iou', root_xpubs: [key.xpub], quorum: 1, definition: {currency: 'USD'})
chain.assets.create(alias: 'bank1_euro_iou', root_xpubs: [key.xpub], quorum: 1, definition: {currency: 'Euro'})
chain.assets.create(alias: 'bank2_usd_iou', root_xpubs: [key.xpub], quorum: 1, definition: {currency: 'USD'})
chain.accounts.create(alias: 'bank1', root_xpubs: [key.xpub], quorum: 1)
chain.accounts.create(alias: 'bank2', root_xpubs: [key.xpub], quorum: 1)

chain.transactions.submit(signer.sign(chain.transactions.build { |b|
  b.issue asset_alias: 'bank1_usd_iou', amount: 2000000
  b.issue asset_alias: 'bank2_usd_iou', amount: 2000000
  b.issue asset_alias: 'bank1_euro_iou', amount: 2000000
  b.control_with_account account_alias: 'bank1', asset_alias: 'bank1_usd_iou', amount: 1000000
  b.control_with_account account_alias: 'bank1', asset_alias: 'bank1_euro_iou', amount: 1000000
  b.control_with_account account_alias: 'bank1', asset_alias: 'bank2_usd_iou', amount: 1000000
  b.control_with_account account_alias: 'bank2', asset_alias: 'bank1_usd_iou', amount: 1000000
  b.control_with_account account_alias: 'bank2', asset_alias: 'bank1_euro_iou', amount: 1000000
  b.control_with_account account_alias: 'bank2', asset_alias: 'bank2_usd_iou', amount: 1000000
}))

# snippet list-alice-transactions
alice_txs = chain.transactions.query(
  filter: 'inputs(account_alias=$1) OR outputs(account_alias=$1)',
  filter_params: ['alice'],
).each do |tx|
  puts "Alice's transaction: #{tx.id}"

  tx.inputs.each do |input|
    puts "-#{input.amount} #{input.asset_alias}"
  end

  tx.outputs.each do |output|
    puts "+#{output.amount} #{output.asset_alias}"
  end
end
# endsnippet

# snippet list-local-transactions
chain.transactions.query(
  filter: 'is_local=$1',
  filter_params: ['yes'],
).each do |tx|
  puts "Local transaction #{tx.id}"
end
# endsnippet

# snippet list-local-assets
chain.assets.query(
  filter: 'is_local=$1',
  filter_params: ['yes'],
).each do |asset|
  puts "Local asset #{asset.id} (#{asset.alias})"
end
# endsnippet

# snippet list-usd-assets
chain.assets.query(
  filter: 'definition.currency=$1',
  filter_params: ['USD'],
).each do |asset|
  puts "USD asset #{asset.id} (#{asset.alias})"
end
# endsnippet

# snippet list-checking-accounts
chain.accounts.query(
  filter: 'tags.type=$1',
  filter_params: ['checking'],
).each do |account|
  puts "Checking account #{account.id} (#{account.alias})"
end
# endsnippet

# snippet list-alice-unspents
chain.unspent_outputs.query(
  filter: 'account_alias=$1',
  filter_params: ['alice'],
).each do |utxo|
  puts "Alice's unspent output: #{utxo.amount} #{utxo.asset_alias}"
end
# endsnippet

# snippet account-balance
chain.balances.query(
  filter: 'account_alias=$1',
  filter_params: ['bank1'],
).each do |b|
  puts "Bank 1 balance of #{b.sum_by['asset_alias']}: #{b.amount}"
end
# endsnippet

# snippet usd-iou-circulation
circulation = chain.balances.query(
  filter: 'asset_alias=$1',
  filter_params: ['bank1_usd_iou'],
).first

puts("Total circulation of Bank 1 USD IOU:  #{circulation.amount}")
# endsnippet

# snippet account-balance-sum-by-currency
chain.balances.query(
  filter: 'account_alias=$1',
  filter_params: ['bank1'],
  sum_by: ['asset_definition.currency']
).each do |b|
  denom = b.sum_by['asset_definition.currency']
  puts "Bank 1 balance of #{denom}-denominated currencies: #{b.amount}"
end
# endsnippet

File: docs/future/core/examples/node/queries.js
const chain = require('chain-sdk')

const client = new chain.Client()
const signer = new chain.HsmSigner()
let key

Promise.all([
  client.mockHsm.keys.create(),
]).then(keys => {
  key  = keys[0].xpub

  signer.addKey(key, client.mockHsm.signerConnection)
}).then(() => Promise.all([
  client.assets.create({
    alias: 'gold',
    rootXpubs: [key],
    quorum: 1
  }),
  client.assets.create({
    alias: 'silver',
    rootXpubs: [key],
    quorum: 1
  }),
  client.accounts.create({
    alias: 'alice',
    tags: {
      type: 'checking'
    },
    rootXpubs: [key],
    quorum: 1
  }),
  client.accounts.create({
    alias: 'bob',
    rootXpubs: [key],
    quorum: 1
  })
])
).then(() =>
  client.transactions.build(builder => {
    builder.issue({ assetAlias: 'gold', amount: 1000 })
    builder.issue({ assetAlias: 'silver', amount: 1000 })
    builder.controlWithAccount({
      accountAlias: 'alice',
      assetAlias: 'gold',
      amount: 1000
    })
    builder.controlWithAccount({
      accountAlias: 'bob',
      assetAlias: 'silver',
      amount: 1000
    })
  }).then(issuance => signer.sign(issuance))
    .then(signed => client.transactions.submit(signed))
).then(() =>
  client.transactions.build(builder => {
    builder.spendFromAccount({
      accountAlias: 'alice',
      assetAlias: 'gold',
      amount: 10
    })
    builder.spendFromAccount({
      accountAlias: 'bob',
      assetAlias: 'silver',
      amount: 10
    })
    builder.controlWithAccount({
      accountAlias: 'alice',
      assetAlias: 'silver',
      amount: 10
    })
    builder.controlWithAccount({
      accountAlias: 'bob',
      assetAlias: 'gold',
      amount: 10
    })
  }).then(trade => signer.sign(trade))
    .then(signed => client.transactions.submit(signed))
).then(() => Promise.all([
  client.assets.create({
    alias: 'bank1UsdIou',
    rootXpubs: [key],
    quorum: 1,
    definition: {
      currency: 'USD'
    }
  }),
  client.assets.create({
    alias: 'bank1EuroIou',
    rootXpubs: [key],
    quorum: 1,
    definition: {
      currency: 'Euro'
    }
  }),
  client.assets.create({
    alias: 'bank2UsdIou',
    rootXpubs: [key],
    quorum: 1,
    definition: {
      currency: 'USD'
    }
  }),
  client.accounts.create({
    alias: 'bank1',
    rootXpubs: [key],
    quorum: 1
  }),
  client.accounts.create({
    alias: 'bank2',
    rootXpubs: [key],
    quorum: 1
  })
])
).then(() =>
  client.transactions.build(builder => {
    builder.issue({ assetAlias: 'bank1UsdIou', amount: 2000000 })
    builder.issue({ assetAlias: 'bank2UsdIou', amount: 2000000 })
    builder.issue({ assetAlias: 'bank1EuroIou', amount: 2000000 })
    builder.controlWithAccount({
      accountAlias: 'bank1',
      assetAlias: 'bank1UsdIou',
      amount: 1000000
    })
    builder.controlWithAccount({
      accountAlias: 'bank1',
      assetAlias: 'bank1EuroIou',
      amount: 1000000
    })
    builder.controlWithAccount({
      accountAlias: 'bank1',
      assetAlias: 'bank2UsdIou',
      amount: 1000000
    })
    builder.controlWithAccount({
      accountAlias: 'bank2',
      assetAlias: 'bank1UsdIou',
      amount: 1000000
    })
    builder.controlWithAccount({
      accountAlias: 'bank2',
      assetAlias: 'bank1EuroIou',
      amount: 1000000
    })
    builder.controlWithAccount({
      accountAlias: 'bank2',
      assetAlias: 'bank2UsdIou',
      amount: 1000000
    })
  }).then(issuance => signer.sign(issuance))
    .then(signed => client.transactions.submit(signed))
).then(() =>

  // snippet list-alice-transactions
  client.transactions.queryAll({
    filter: 'inputs(account_alias=$1) OR outputs(account_alias=$1)',
    filterParams: ['alice'],
  }, (tx, next, done) => {
    console.log("Alice's transaction: " + tx.id)

    tx.inputs.forEach(input => {
      console.log('-' + input.amount + ' ' + input.assetAlias)
    })

    tx.outputs.forEach(output => {
      console.log('+' + output.amount + ' ' + output.assetAlias)
    })

    // next() moves to the next item.
    // done() terminates the loop early, and causes the
    //   query promise to resolve. Passing an error will reject
    //   the promise.
    next()
  })
  // endsnippet

).then(() =>

  // snippet list-checking-transactions
  client.transactions.queryAll({
    filter: 'inputs(account_tags.type=$1) OR outputs(account_tags.type=$1)',
    filterParams: ['checking'],
  }, (tx, next, done) => {
    console.log("Checking account transaction: " + tx.id)

    tx.inputs.forEach(input => {
      console.log('-' + input.amount + ' ' + input.assetAlias)
    })

    tx.outputs.forEach(output => {
      console.log('+' + output.amount + ' ' + output.assetAlias)
    })

    // next() moves to the next item.
    // done() terminates the loop early, and causes the
    //   query promise to resolve. Passing an error will reject
    //   the promise.
    next()
  })
  // endsnippet

).then(() =>

  // snippet list-local-transactions
  client.transactions.queryAll({
    filter: 'is_local=$1',
    filterParams: ['yes']
  }, (tx, next, done) => {
    console.log('Local transaction ' + tx.id)

    // next() moves to the next item.
    // done() terminates the loop early, and causes the
    //   query promise to resolve. Passing an error will reject
    //   the promise.
    next()
  })
  // endsnippet

).then(() =>

  // snippet list-local-assets
  client.assets.queryAll({
    filter: 'is_local=$1',
    filterParams: ['yes']
  }, (asset, next, done) => {
    console.log('Local asset ' + asset.id + ' (' + asset.alias + ')')

    // next() moves to the next item.
    // done() terminates the loop early, and causes the
    //   query promise to resolve. Passing an error will reject
    //   the promise.
    next()
  })
  // endsnippet

).then(() =>

  // snippet list-usd-assets
  client.assets.queryAll({
    filter: 'definition.currency=$1',
    filterParams: ['USD']
  }, (asset, next, done) => {
    console.log('USD asset ' + asset.id + ' (' + asset.alias + ')')

    // next() moves to the next item.
    // done() terminates the loop early, and causes the
    //   query promise to resolve. Passing an error will reject
    //   the promise.
    next()
  })
  // endsnippet

).then(() =>

  // snippet list-checking-accounts
  client.accounts.queryAll({
    filter: 'tags.type=$1',
    filterParams: ['checking']
  }, (account, next, done) => {
    console.log('Checking account ' + account.id + ' (' + account.alias + ')')

    // next() moves to the next item.
    // done() terminates the loop early, and causes the
    //   query promise to resolve. Passing an error will reject
    //   the promise.
    next()
  })
  // endsnippet

).then(() =>

  // snippet list-alice-unspents
  client.unspentOutputs.queryAll({
    filter: 'account_alias=$1',
    filterParams: ['alice']
  }, (utxo, next, done) => {
    console.log("Alice's unspent output: " + utxo.amount + ' ' + utxo.assetAlias)

    // next() moves to the next item.
    // done() terminates the loop early, and causes the
    //   query promise to resolve. Passing an error will reject
    //   the promise.
    next()
  })
  // endsnippet

).then(() =>

  // snippet list-checking-unspents
  client.unspentOutputs.queryAll({
    filter: 'account_tags.type=$1',
    filterParams: ['checking']
  }, (utxo, next, done) => {
    console.log("Checking account unspent output: " + utxo.amount + ' ' + utxo.assetAlias)

    // next() moves to the next item.
    // done() terminates the loop early, and causes the
    //   query promise to resolve. Passing an error will reject
    //   the promise.
    next()
  })
  // endsnippet

).then(() =>

  // snippet account-balance
  client.balances.queryAll({
    filter: 'account_alias=$1',
    filterParams: ['bank1']
  }, (balance, next, done) => {
    console.log('Bank 1 balance of ' + balance.sumBy.assetAlias + ': ' + balance.amount)

    // next() moves to the next item.
    // done() terminates the loop early, and causes the
    //   query promise to resolve. Passing an error will reject
    //   the promise.
    next()
  })
  // endsnippet

).then(() =>

  // snippet checking-accounts-balance
  client.balances.queryAll({
    filter: 'account_tags.type=$1',
    filterParams: ['checking']
  }, (balance, next, done) => {
    console.log('Checking accounts balance of ' + balance.sumBy.assetAlias + ': ' + balance.amount)

    // next() moves to the next item.
    // done() terminates the loop early, and causes the
    //   query promise to resolve. Passing an error will reject
    //   the promise.
    next()
  })
  // endsnippet

).then(() =>

  // snippet usd-iou-circulation
  client.balances.queryAll({
    filter: 'asset_alias=$1',
    filterParams: ['bank1UsdIou']
  }, (balance, next, done) => {
    console.log('Total circulation of Bank 1 USD IOU: ' + balance.amount)

    // next() moves to the next item.
    // done() terminates the loop early, and causes the
    //   query promise to resolve. Passing an error will reject
    //   the promise.
    next()
  })
  // endsnippet

).then(() =>

  // snippet account-balance-sum-by-currency
  client.balances.queryAll({
    filter: 'account_alias=$1',
    filterParams: ['bank1'],
    sumBy: ['asset_definition.currency']
  }, (balance, next, done) => {
    var denom = balance.sumBy['assetDefinition.currency']
    console.log('Bank 1 balance of ' + denom + '-denominated currencies: ' + balance.amount)

    // next() moves to the next item.
    // done() terminates the loop early, and causes the
    //   query promise to resolve. Passing an error will reject
    //   the promise.
    next()
  })
  // endsnippet

).catch(err =>
  process.nextTick(() => { throw err })
)
